lua/wire/flir.lua

```diff
@@ -37,126 +56,160 @@ if CLIENT then
 		["$pp_colour_brightness"] = 0
 	}
 
-	local function SetFLIRMat(ent)
-		if not IsValid(ent) then return end
+	--add and remove entities from the FLIR rendering stack
 
-		if (ent:GetMoveType() == MOVETYPE_VPHYSICS or ent:IsPlayer() or ent:IsNPC() or ent:IsRagdoll() or ent:GetClass() == "gmod_wire_hologram") and ent:GetColor().a > 0 then
-			ent.RenderOverride = FLIR.Render
-			FLIR.RenderStack[ent] = true
-		end
+	local function RemoveFLIR(ent)
+		FLIR.RenderStack[ent] = nil
+		if IsValid(ent) then ent.RenderOverride = nil end
 	end
 
-	local function RemoveFLIRMat(ent)
-		ent.RenderOverride = nil
-		FLIR.RenderStack[ent] = nil
+	local function RemoveCheck(ent)		--make sure entity has actually been removed, not weird call
+		timer.Create(tostring(ent).."rmcheck", 0, 1, function()
+			if not IsValid(ent) then RemoveFLIR(ent) end
+		end)
 	end
 
-	function FLIR.Render(self)
-		if FLIR.ShouldRender then self:DrawModel() end
+	local function SetFLIR(ent)
+		if not IsValid(ent) then return end
+		local c = ent:GetClass()
+
+		if (c == "prop_physics" or ent:GetMoveType() == MOVETYPE_VPHYSICS or (ent:IsPlayer() and ent != ply) or ent:IsNPC() or ent:IsRagdoll() or c == "gmod_wire_hologram") and ent:GetColor().a > 0 then
+			if ent:GetColor().a > 0 then
+				FLIR.RenderStack[ent] = true
+				ent.RenderOverride = FLIR.Render	--we're already rendering later, so don't bother beforehand
+				ent:CallOnRemove("rmcheck", RemoveCheck(ent))
```

Comment:
Remove the `(ent)`

Reply:
Better yet, replace with
```lua
ent:CallOnRemove("rmcheck", function()
	timer.Simple(0, function()
		if not ent:IsValid() then RemoveFLIR(ent) end
	end)
end)
```

Reply:
Is this even needed? The loop on line 118 will remove it from flir anyway.

Reply:
Yeah honestly I see your point, it's a leftover from before I did the loop stuff. will fix.